job-runner/chain/chain.go

```diff
@@ -145,19 +122,24 @@ func (c *Chain) IsDoneRunning() (done bool, complete bool) {
 		case proto.STATE_RUNNING:
 			// If any jobs are still running, the chain isn't done or complete.
 			return false, false
-		case proto.STATE_PENDING, proto.STATE_STOPPED:
-			// If any job can be run, the chain is not done or complete.
-			// Treat stopped jobs as pending jobs because they may be retried,
-			// as when resuming a suspended job chain.
+		case proto.STATE_STOPPED:
```

Comment:
I'm a little confused about this. When we suspend job chains, there could be many STOPPED jobs, correct? Then when we go to resume the job chain, won't there be a lot of STOPPED jobs in it that we rerun (i.e., there will be STOPPED jobs that are runnable)? Or do those get immediately switched to PENDING when a suspended job chain is resumed or something?

Is this scenario possible? - we suspend a job chain that results in 2 stopped jobs. Then we go to resume the job chain, which starts running those 2 jobs. The first job completes successfully before the next job has even started running, at which point if this function is called, it will return done=true, even though there's still work to be done on the current chain run.

Reply:
First paragraph addressed in code comments for this func:
```go
// Stopped jobs are not runnable in this context (i.e. chain context). This
// function applies to the current chain run. Once a job is stopped, it cannot
// be re-ran in the current chain run. If the chain is re-ran (i.e. resumed),
// IsRunnable will return true for stopped jobs because stopped jobs are runnable
// in that context (i.e. job context).
```
In "chain context", stopped == done. But in "job context", stopped == (re)runnable. And yes, the stopped jobs are switched to state RUNNING (not PENDING).

Second paragraph: hm, yes I think that could happen. If job1 finishes and is reaped very quickly, after reaping the reaper calls chain.IsDoneRunning which will see job2 as state stopped because its state wasn't changed in runJobs yet. Solution: on SJC resume, set all stopped jobs to pending very first thing.

Reply:
Done. As we discussed in Slack: the traverser should be given a ready-to-run chain, so it stays free of any such logic. This logic is now in `MakeFromSJC` (traverser factory). Currently, we have only SuspendedJobChain which needs to be fixed up to become a normal, read-to-run JobChain, a d in future if we have other chain types, some higher-level code responsible for those chain types can fix them up before running in a traverser (else traverser would get cluttered with logic for all the different chain types).